ios/brave-ios/Sources/Brave/Helpers/URLBarHelper.swift

```diff
@@ -0,0 +1,1322 @@
+        if pp < probHashThreshold * 1.5 {
+          return true
+        }
+      }
+    }
+
+    return false
+  }
+
+  private func checkForLongNumber(_ str: String, _ maxNumberLength: Int) -> String? {
+    let controlString = str.replacingOccurrences(
+      of: "[^A-Za-z0-9]",
+      with: "",
+      options: .regularExpression
+    )
+
+    var location = 0
+    var maxLocation = 0
+    var maxLocationPosition: String.Index? = nil
+
+    for i in controlString.indices {
+      if controlString[i] >= "0" && controlString[i] <= "9" {
+        location += 1
+      } else {
+        if location > maxLocation {
+          maxLocation = location
+          maxLocationPosition = i
+        }
+
+        location = 0
+      }
+    }
+
+    if location > maxLocation {
+      maxLocation = location
+      maxLocationPosition = controlString.endIndex
+    }
+
+    if let maxLocationPosition = maxLocationPosition, maxLocation > maxNumberLength {
+      let start = controlString.index(maxLocationPosition, offsetBy: -maxLocation)
+      let end = maxLocationPosition
+
+      return String(controlString[start..<end])
+    } else {
+      return nil
+    }
+  }
+
+  private func checkForEmail(_ str: String) -> Bool {
+    let emailPattern = "[a-z0-9\\-_@]+(@|%40|%(25)+40)[a-z0-9\\-_]+\\.[a-z0-9\\-_]"
+
+    guard
+      let emailRegex = try? NSRegularExpression(pattern: emailPattern, options: .caseInsensitive)
+    else {
+      return false
+    }
+
+    let range = NSRange(location: 0, length: str.utf16.count)
+    return emailRegex.firstMatch(in: str, options: [], range: range) != nil
+  }
+
+  private func isHashProb(_ str: String) -> Double {
+    var logProb = 0.0
+    var transC = 0
+    let filteredStr = str.replacingOccurrences(
+      of: "[^A-Za-z0-9]",
+      with: "",
+      options: .regularExpression
+    )
+
+    let characters = Array(filteredStr)
+    for i in 0..<(characters.count - 1) {
+      if let pos1 = probHashChars[characters[i]], let pos2 = probHashChars[characters[i + 1]] {
+        logProb += probHashLogM[pos1][pos2]
+        transC += 1
+      }
+    }
+
+    if transC > 0 {
+      return exp(logProb / Double(transC))
+    } else {
+      return exp(logProb)
+    }
+  }
+
+  private let probHashThreshold: Double = 0.015
+
+  private let probHashChars: [Character: Int] = [
+    "1": 1, "0": 0, "3": 3, "2": 2, "5": 5, "4": 4, "7": 7, "6": 6, "9": 9, "8": 8,
+    "A": 36, "C": 38, "B": 37, "E": 40, "D": 39, "G": 42, "F": 41, "I": 44, "H": 43,
+    "K": 46, "J": 45, "M": 48, "L": 47, "O": 50, "N": 49, "Q": 52, "P": 51, "S": 54,
+    "R": 53, "U": 56, "T": 55, "W": 58, "V": 57, "Y": 60, "X": 59, "Z": 61, "a": 10,
+    "c": 12, "b": 11, "e": 14, "d": 13, "g": 16, "f": 15, "i": 18, "h": 17, "k": 20,
+    "j": 19, "m": 22, "l": 21, "o": 24, "n": 23, "q": 26, "p": 25, "s": 28, "r": 27,
+    "u": 30, "t": 29, "w": 32, "v": 31, "y": 34, "x": 33, "z": 35,
+  ]
+
+  private let probHashLogM: [[Double]] = [
```

Comment:
Where do these numbers come from?

Reply:
So the `isSuspiciousQuery` method https://github.com/brave/search/blob/bee821698102679695a9b3bfafb8abef6b595d81/services/mixer/serpkit/src/lib/utils/hw.ts#L376 mentioned in the description uses 

isHashProb method as last step for checking the hash evaluationhttps://github.com/brave/search/blob/bee821698102679695a9b3bfafb8abef6b595d81/services/mixer/serpkit/src/lib/utils/hw.ts#L15

and this method is using values from data file which is 
https://github.com/brave/search/blob/bee821698102679695a9b3bfafb8abef6b595d81/services/mixer/serpkit/src/lib/utils/data.ts#L5

so we had to copy pasta 🍝 3 values probHashThreshold, probHashChars, probHashLogM

Reply:
Moved to a different file at https://github.com/brave/brave-core/pull/24376/commits/399c737568781bca5bde7107be23640fffdab797